
<robot name="underwater" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- * * * Color Definitions * * * -->
    

    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>

    <material name="green2">
        <color rgba="0.8 0.9 0.8 0.9"/>
    </material> 

    <material name="red">
        <color rgba="0.8 0 0 1"/>
    </material>

    <material name="gray">
        <color rgba="0.5 0.5 0.5 0.9"/>
    </material>

    <material name="transparent">
        <color rgba="1 1 1 0.4"/>
    </material>

    <material name="full_transparent">
        <color rgba="1 1 1 0"/>
    </material>

    <!-- * * * Link Definitions * * * -->

    <!-- mesh files -->
    <link name="robot_base_link">
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="5.99" />
            <inertia ixx="0.08319857" ixy="0.0008657224" ixz="0.0009579166" iyy="0.08357262" iyz="0.0001832366" izz="0.12186"/>
        </inertial>
        
        <collision>
            <origin xyz="0 0 .01" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.25" length="0.11"/>
            </geometry>
        </collision>

        <visual>
            <origin rpy="0 0 0" xyz="0 0 -.0976"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/MainBody.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
        <plugin name="thruster_plug" filename="libthruster_plug.so"/>
    </link> 

    <link name="center_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.0" />
            <inertia ixx="0" ixy="0.0" ixz="0.0" iyy="0" iyz="0.0" izz="0"/>
        </inertial>
        <visual>
            <origin rpy="0 0 0" xyz="-0.0302 -0.03215 -0.018"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/CentralPropeller.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="black"/>
        </visual>
    </link> 


    <link name="shaft1_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link> 

    <link name="shaft2_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>  

    <link name="shaft3_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link> 

    <link name="cylinder1_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
        
    </link> 
    <link name="cylinder2_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link>

    <link name="cylinder3_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0  -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="motorcylinder1_link">

        <visual>
            <origin rpy="0 0 0" xyz="0 .244 0.00"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="motorcylinder2_link">

        <visual>
            <origin rpy="0 0 -2.094" xyz=".2114 -.122 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="motorcylinder3_link">

        <visual>
            <origin rpy="0 0 2.094" xyz="-.2114 -.122 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 


    <link name="wheel1_link">
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.361" />
            <inertia ixx="0.0007561942" ixy="0.0" ixz="0.0" iyy="0.001441249" iyz="0.0" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="wheel1_link">
        <kp>100000.0</kp>
        <kd>1.0</kd>
        <mu1>5.0</mu1>
        <mu2>.1</mu2>
        <fdir1>0 1 0</fdir1>
        <material>Gazebo/Green</material>
    </gazebo>
    <link name="wheel2_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.361" />
            <inertia ixx="0.0007561942" ixy="0.0" ixz="0.0" iyy="0.001441249" iyz="0.0" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="wheel2_link">
        <kp>100000.0</kp>
        <kd>1.0</kd>
        <mu1>5.00</mu1>
        <mu2>.1</mu2>
        <fdir1>0 1 0</fdir1>    
        <material>Gazebo/Green</material>
    </gazebo>


    <link name="wheel3_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.361" />
            <inertia ixx="0.0007561942" ixy="0.0" ixz="0.0" iyy="0.001441249" iyz="0.0" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>
    <gazebo reference="wheel3_link">
        <kp>100000.0</kp>
        <kd>1.0</kd>
        <mu1>5.0</mu1>
        <mu2>.1</mu2>
        <fdir1>0 1 0</fdir1> 
        <material>Gazebo/Green</material>
    </gazebo> 
    
    <!-- * * * Joint Definitions * * * -->

    <joint name="center_motor" type="continuous">
        <parent link="robot_base_link"/>
        <child link="center_link"/>
        <origin xyz="0.000 0 0.0000" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="motor1" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft1_link"/>
        <origin xyz="0 0.245 0" rpy="0 0 0"/>
        <axis xyz="0 1.5708 0"/>
        <dynamics damping="0.2"/>
    </joint>  

    <joint name="motor2" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft2_link"/>
        <origin xyz="-.21223 -0.1224 0" rpy="0 0 2.094"/>
        <axis xyz="0 1 0 "/>
        <dynamics damping="0.2"/>
    </joint> 
  
    <joint name="motor3" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft3_link"/>
        <origin xyz=".21223 -0.1224 0" rpy="0 0 -2.094"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.2"/>
    </joint>  

    <joint name="wheel1_joint" type="fixed">
        <parent link="shaft1_link"/>
        <child link="wheel1_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
   </joint>

         
    <joint name="wheel2_joint" type="fixed">
        <parent link="shaft2_link"/>
        <child link="wheel2_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
   </joint>

    <joint name="wheel3_joint" type="fixed">
        <parent link="shaft3_link"/>
        <child link="wheel3_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder1_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder1_link"/>
        <origin rpy="0 0 0" xyz=".1048 .0605 -.00"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder2_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder2_link"/>
        <origin rpy="0 0 0" xyz="-.1048 .0605 -.00"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder3_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder3_link"/>
        <origin rpy="0 0 0" xyz="0 -.121 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="motorcylinder1_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder1_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="motorcylinder2_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder2_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="motorcylinder3_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder3_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>
    <transmission name="motor1_tran">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="$motor1_act">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
        <joint name="motor1">
            <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>
    <transmission name="motor2_tran">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="$motor2_act">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
        <joint name="motor2">
            <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>
    <transmission name="motor3_tran">
        <type>transmission_interface/SimpleTransmission</type>
        <actuator name="$motor3_act">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
        <joint name="motor3">
            <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
    </transmission>

    <gazebo>
     <plugin name="thruster_plug" filename="libthruster_plug.so">
     <BodyName>robot_base_link</BodyName>
    </plugin>
    </gazebo>








    <gazebo reference="cylinder1_link">
        <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="cylinder2_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="cylinder3_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder1_link">
        <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder2_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder3_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
        <robotNamespace>/UGazebo</robotNamespace>
    </plugin>
    </gazebo>

</robot> 
